tests/conditional_ksample/test_cd.py

```diff
@@ -0,0 +1,233 @@
+        cd_func.condind(
+            X=X,
+            Y=Y,
+            group_ind=group_ind,
+            propensity_weights=np.ones((100, 2)) * 0.5,
+            random_seed=seed,
+        )
+
+
+@pytest.mark.parametrize(
+    ["cd_func", "cd_kwargs"],
+    [
+        [bregman, dict()],
+        [kcd, dict()],
+        [bregman, {"propensity_model": RandomForestClassifier(n_estimators=50, random_state=seed)}],
+        [bregman, {"propensity_weights": np.ones((n_samples, 2)) * 0.5}],
+        [kcd, {"propensity_model": RandomForestClassifier(n_estimators=50, random_state=seed)}],
+        [kcd, {"propensity_weights": np.ones((n_samples, 2)) * 0.5}],
+    ],
+)
+def test_given_single_environment_when_conditional_ksample_fail_to_reject_null(cd_func, cd_kwargs):
+    """Test conditional discrepancy tests on a single environment.
+
+    Here the distributions of each variable are the same across groups of data.
+    """
+    df = _single_env_scm(n_samples=n_samples, offset=2.0, seed=rng)
+
+    group_col = "group"
+    alpha = 0.1
+    null_sample_size = 25
+
+    res = cd_func.condind(
+        X=df["x"],
+        Y=df["x1"],
+        group_ind=df[group_col],
+        null_sample_size=null_sample_size,
+        **cd_kwargs,
+        random_seed=seed,
+    )
+    assert res.pvalue > alpha, f"Fails with {res.pvalue} not greater than {alpha}"
+    res = cd_func.condind(
+        X=df["x"],
+        Y=df["z"],
+        group_ind=df[group_col],
+        null_sample_size=null_sample_size,
+        n_jobs=-1,
+        **cd_kwargs,
+        random_seed=seed,
+    )
+    assert res.pvalue > alpha, f"Fails with {res.pvalue} not greater than {alpha}"
+    res = cd_func.condind(
+        X=df["x"],
+        Y=df["y"],
+        group_ind=df[group_col],
+        null_sample_size=null_sample_size,
+        n_jobs=-1,
+        **cd_kwargs,
+        random_seed=seed,
+    )
+    assert res.pvalue > alpha, f"Fails with {res.pvalue} not greater than {alpha}"
+
+
+@pytest.mark.parametrize(
+    ["cd_func", "cd_kwargs"],
+    [
+        [bregman, dict()],
+        [kcd, dict()],
+        [bregman, {"propensity_model": RandomForestClassifier(n_estimators=50, random_state=seed)}],
+        [kcd, {"propensity_model": RandomForestClassifier(n_estimators=50, random_state=seed)}],
+    ],
+)
+def test_cd_simulation_multi_environment(cd_func, cd_kwargs):
+    """Test conditional discrepancy tests with multiple environments.
+
+    In this setting, we form a selection diagram
+
+    X1 -> Y <- X and Y -> Z, with S-nodes pointing to X1 and X. That means
+    the distributions of X1 and X are different across groups. Therefore
+    we should reject the null hypothesis, except for P(Z | Y).
+    """
+    df = _multi_env_scm(n_samples=n_samples // 2, offset=2.0, seed=rng)
+
+    group_col = "group"
+    alpha = 0.1
+    null_sample_size = 20
+
+    res = cd_func.condind(
+        X=df[["x"]],
+        Y=df[["z"]].copy(),
```

Comment:
Why do we need `copy` here?

Reply:
Good question. I don't think we do. I will remove.

Reply:
Removed.